### PR TITLE
New Example for requesting JSON data

### PR DESCRIPTION
**Pull request by agnosticdev**
**Motivation:**

Provide a starter example of how to process HTTP requests for JSON data.
    
**Modifications:**
    
Added a new example called APIGatewayEndpoint.
This sample is deployed with the template: APIGatewayEndpoint-template.yml
    
**Result:**
A sample API to request JSON data.

```bash
# (200 - All Articles)
% curl -v https://{baseurl}/articles

# (200 - Article 1)
% curl -v https://{baseurl}/articles?aid=1

# (200 - Article 2)
% curl -v https://{baseurl}/articles?aid=2

# (200 - Article 3)
% curl -v https://{baseurl}/articles?aid=3

# (200 - No articles found)
% curl -v https://{baseurl}/articles?aid=4  

# (404 - Not Found)
% curl -v https://{baseurl}/articl
```

### CODE REVIEW

File: Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        let body2 = "All the Lorem Ipsum generators on the Internet tend to repeat predefined chunks as necessary, making this the first true generator on the Internet."
+        let body3 = "If you are going to use a passage of Lorem Ipsum, you need to be sure there isn't anything embarrassing hidden in the middle of text."
+        let article1 = Article(aid: 1, title: "Article 1", body: body1, date: Date(), comments: [])
+        let article2 = Article(aid: 2, title: "Article 2", body: body2, date: Date(), comments: [])
+        let article3 = Article(aid: 4, title: "Article 3", body: body3, date: Date(), comments: [])
+        articles.append(article1)
+        articles.append(article2)
+        articles.append(article3)
+    }
+}
```
Comment (fabianfett) [inline Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift:76]: We could maybe move this to a separate file?

File: Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+                
+                context.logger.info("[API Gateway Endpoint] - articles count: \(articles.count)")
+                let jsonEncodedArticles = try encoder.encode(articleSubSet)
+                let jsonArticles = String(data: jsonEncodedArticles, encoding: .utf8)!
+                response.0 = 200
+                response.1 = jsonArticles
+                context.logger.info("[API Gateway Endpoint] - JSON articles decoded")
+            } catch {
+                response.0 = 500
+                context.logger.info("[API Gateway Endpoint] - Error encoding articles")
```
Comment (fabianfett) [inline Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift:131]: in this case we should probably use the `error` log level. further we want probably be able to read the error.

File: Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+struct APIGatewayProxyLambda: EventLoopLambdaHandler {
+    public typealias In = APIGateway.V2.Request
+    public typealias Out = APIGateway.V2.Response
+
+    public func handle(context: Lambda.Context, event: APIGateway.V2.Request) -> EventLoopFuture<APIGateway.V2.Response> {
+        context.logger.info("[API Gateway Endpoint] Response handler")
+        // Load the test data
+        // This is where a DynamoDB table could be monitored to pull in articles instead of loading them locally.
+        let testData = TestData()
```
[the remark this answers is not in the document]
Comment (agnosticdev) [reply Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift]: I moved this out to a property on `APIGatewayProxyLambda`

File: Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                response.0 = 200
+                response.1 = jsonArticles
+                context.logger.info("[API Gateway Endpoint] - JSON articles decoded")
+            } catch {
+                response.0 = 500
+                context.logger.info("[API Gateway Endpoint] - Error encoding articles")
+                response.1 = error.localizedDescription
+            }
+            return response
+        }
```
[the remark this answers is not in the document]
Comment (agnosticdev) [reply Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift]: I moved this out to a private method on `APIGatewayProxyLambda`

File: Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift (hunk 1 above)
Comment (agnosticdev) [reply Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift]: Good idea. Add a new file called `TestData.swift` for this.

File: Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift (hunk 2 above)
Comment (agnosticdev) [reply Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift]: Good catch.  Added `error` as the log level.

File: Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    public typealias In = APIGateway.V2.Request
+    public typealias Out = APIGateway.V2.Response
+
+    public func handle(context: Lambda.Context, event: APIGateway.V2.Request) -> EventLoopFuture<APIGateway.V2.Response> {
+        context.logger.info("[API Gateway Endpoint] Response handler")
+        // Load the test data
+        // This is where a DynamoDB table could be monitored to pull in articles instead of loading them locally.
+        let testData = TestData()
+        // Create an JSONEncoder for encoding a JSON response
+        let encoder = JSONEncoder()
```
[the remark this answers is not in the document]
Comment (agnosticdev) [reply Examples/LambdaFunctions/Sources/APIGatewayEndpoint/main.swift]: Good call.  I wasn't aware the encoder was there. I switched over to this encoder in the latest commit.

**Comment by agnosticdev** [comment]
Updated the PR based on the provided feedback.

**Pull request closed without merging**